File: backend/src/graphql/resolvers/cart.resolver.ts
// Cart GraphQL Resolvers

import { CartService } from '../../services/cart.service.js';
import { GraphQLContext } from '../context.js';
import { GraphQLError } from 'graphql';

export const cartResolvers = {
  Query: {
    cart: async (_: any, __: any, context: GraphQLContext) => {
      if (!context.user) {
        throw new GraphQLError('Not authenticated', {
          extensions: { code: 'UNAUTHENTICATED' },
        });
      }

      return CartService.getCart(context.user.id);
    },
  },

  Mutation: {
    addToCart: async (
      _: any,
      args: { productId: string; quantity: number },
      context: GraphQLContext
    ) => {
      if (!context.user) {
        throw new GraphQLError('Not authenticated', {
          extensions: { code: 'UNAUTHENTICATED' },
        });
      }

      try {
        return await CartService.addToCart(context.user.id, args.productId, args.quantity);
      } catch (error) {
        throw new GraphQLError(error instanceof Error ? error.message : 'Failed to add to cart', {
          extensions: { code: 'BAD_REQUEST' },
        });
      }
    },

    updateCartItem: async (
      _: any,
      args: { itemId: string; quantity: number },
      context: GraphQLContext
    ) => {
      if (!context.user) {
        throw new GraphQLError('Not authenticated', {
          extensions: { code: 'UNAUTHENTICATED' },
        });
      }

      try {
        return await CartService.updateCartItem(context.user.id, args.itemId, args.quantity);
      } catch (error) {
        throw new GraphQLError(error instanceof Error ? error.message : 'Failed to update cart item', {
          extensions: { code: 'BAD_REQUEST' },
        });
      }
    },

    removeFromCart: async (
      _: any,
      args: { itemId: string },
      context: GraphQLContext
    ) => {
      if (!context.user) {
        throw new GraphQLError('Not authenticated', {
          extensions: { code: 'UNAUTHENTICATED' },
        });
      }

      try {
        await CartService.removeFromCart(context.user.id, args.itemId);
        return true;
      } catch (error) {
        throw new GraphQLError(error instanceof Error ? error.message : 'Failed to remove from cart', {
          extensions: { code: 'BAD_REQUEST' },
        });
      }
    },

    clearCart: async (_: any, __: any, context: GraphQLContext) => {
      if (!context.user) {
        throw new GraphQLError('Not authenticated', {
          extensions: { code: 'UNAUTHENTICATED' },
        });
      }

      try {
        await CartService.clearCart(context.user.id);
        return true;
      } catch (error) {
        return false;
      }
    },
  },
};

File: backend-java/src/main/java/com/ecommerce/mini/graphql/CartResolver.java
package com.ecommerce.mini.graphql;

import com.ecommerce.mini.dto.AddToCartRequest;
import com.ecommerce.mini.dto.CartItemResponse;
import com.ecommerce.mini.dto.UpdateCartItemRequest;
import com.ecommerce.mini.service.CartService;
import lombok.RequiredArgsConstructor;
import org.springframework.graphql.data.method.annotation.Argument;
import org.springframework.graphql.data.method.annotation.MutationMapping;
import org.springframework.graphql.data.method.annotation.QueryMapping;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.stereotype.Controller;

import java.util.List;
import java.util.Map;

/**
 * GraphQL Resolver for Cart operations
 */
@Controller
@RequiredArgsConstructor
public class CartResolver {

    private final CartService cartService;

    @QueryMapping
    public List<CartItemResponse> cart() {
        String userId = getCurrentUserId();
        return cartService.getCartItems(userId);
    }

    @MutationMapping
    public CartItemResponse addToCart(@Argument Map<String, Object> input) {
        String userId = getCurrentUserId();

        AddToCartRequest request = new AddToCartRequest();
        request.setProductId((String) input.get("productId"));
        request.setQuantity((Integer) input.get("quantity"));

        return cartService.addToCart(userId, request);
    }

    @MutationMapping
    public CartItemResponse updateCartItem(
            @Argument String itemId,
            @Argument Map<String, Object> input) {

        String userId = getCurrentUserId();

        UpdateCartItemRequest request = new UpdateCartItemRequest();
        request.setQuantity((Integer) input.get("quantity"));

        return cartService.updateCartItem(userId, itemId, request);
    }

    @MutationMapping
    public Boolean removeFromCart(@Argument String itemId) {
        String userId = getCurrentUserId();
        cartService.removeFromCart(userId, itemId);
        return true;
    }

    @MutationMapping
    public Boolean clearCart() {
        String userId = getCurrentUserId();
        cartService.clearCart(userId);
        return true;
    }

    private String getCurrentUserId() {
        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
        if (authentication == null || !authentication.isAuthenticated()) {
            throw new RuntimeException("User not authenticated");
        }
        // In a real scenario, you'd extract the user ID from the JWT token
        // For now, we'll use the email as identifier
        return authentication.getName();
    }
}
